### PR TITLE
feat(jira): return the created issue, sanitize labels and add bulk status lookup

### PR DESCRIPTION
**Pull request by danibarranqueroo**
### Context

Jira issues created through the Prowler Cloud integration have no machine-readable reference back to the finding, and Prowler keeps no durable reference to the created issue (#12448, PROWLER-2432).

This PR provides the SDK-only foundation for the Jira findings-reference epic (PROWLER-2431). The API installs Prowler SDK from `master` through `api/uv.lock`, so the API metadata, deduplication, and UI work must consume these contracts after this PR merges.

#### Creation results

`Jira.send_finding()` returns an immutable `JiraCreationResult` with one of these outcomes:

- `confirmed_success`: Jira returned `201` with a valid issue key, immutable issue ID, and derived browse URL.
- `confirmed_rejection`: Jira definitively rejected the request, credentials, project, or issue type.
- `retryable_failure`: Jira did not receive the request, or the request can be safely retried, such as `429` or `ConnectTimeout`.
- `uncertain`: Jira may have created the issue, such as after a read timeout, ambiguous connection loss, `5xx`, malformed `201`, or incomplete successful response.

The result includes the issue key, immutable ID, browse URL, delivery marker, HTTP status, raw `Retry-After`, and safe error details.

`JiraCreationResult.__bool__()` returns `True` only for `confirmed_success`. Existing callers using `if result:` remain compatible. Explicit API outcome handling and persistence belong to #12541.

The individual and legacy bulk creation paths share the same response classifier.

#### Labels and delivery markers

`Jira.sanitize_label()` preserves case, removes control characters, converts whitespace runs to `_`, collapses repeated `_`, trims edge `_`, and keeps other printable characters.

Stable finding labels use `prowler-finding-`. Delivery-attempt labels use `prowler-attempt-`. When a complete prefixed label exceeds Jira's 255-character limit, the SDK keeps a readable prefix and appends the full SHA-256 hash of the raw value.

`send_finding()` accepts a caller-owned `delivery_attempt_marker`. The SDK includes it in the Jira labels and never replaces it during retries.

#### Delivery-marker lookup

`Jira.search_issues_by_delivery_attempt()` uses Jira's enhanced JQL search and returns typed search results:

- Successful lookup, including zero matches.
- Retryable lookup failure.
- Unknown lookup result.

Every match includes the immutable issue ID, current key, and browse URL. Repeated pagination tokens return `unknown` instead of allowing an unbounded request loop.

#### Issue-status lookup

`Jira.get_issues_status()` accepts `JiraIssueReference` values containing the immutable issue ID and stored key. It returns one typed result per unique input in first-seen order.

Lookups are sent by immutable ID in batches of 100 and return:

- `open`
- `done`
- `moved`
- `missing`
- `forbidden`
- `unknown`

`moved` takes precedence when the immutable ID resolves to a different key. Done status uses `statusCategory.key == "done"` instead of the customizable display name.

Per-issue `403` and `404` observations become `forbidden` and `missing`. Omitted, malformed, or failed results become `unknown`. Successful batches are preserved if a later batch fails. `missing` never means confirmed deletion.

#### Destination validation

Validated `(project_key, issue_type)` destinations are cached per Jira instance. An entry is stored only after both catalog checks pass, avoiding repeated project and issue-type requests for multiple findings sent to the same destination.

This PR has no API source changes, database state, retry ledger, migrations, or UI changes. Those changes remain in the dependent PRs.

### Steps to review

1. Run the Jira SDK tests:

   `uv run pytest tests/lib/outputs/jira/jira_test.py -q`

   Expected result: `156 passed`.

2. Verify the current API caller remains compatible:

   `cd api && uv run pytest src/backend/tasks/tests/test_integrations.py -k TestJiraIntegration -q`

   Expected result: `7 passed`.

3. Confirm there is no API diff in this SDK-only PR:

   `git diff origin/master...HEAD -- api`

4. Check the final diff:

   `git diff --check`

### Checklist

<details>

<summary><b>Community Checklist</b></summary>

- [x] This feature/issue is listed in the [open issues](https://github.com/prowler-cloud/prowler/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen) or roadmap.prowler.com
- [x] Is it assigned to me, if not, request it through the [open issues](https://github.com/prowler-cloud/prowler/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen) or [Prowler Community Slack](https://goto.prowler.com/slack)
- [x] I reviewed the [open pull requests](https://github.com/prowler-cloud/prowler/pulls?q=sort%3Aupdated-desc+is%3Apr+is%3Aopen) and confirmed there is no existing PR that implements the same outcome

</details>

- [x] Review if the code is covered by tests.
- [x] Review if the code is documented following https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings
- [ ] Review if a backport is needed.
- [ ] Review if the README needs changes.
- [x] Ensure a changelog fragment is added under `prowler/changelog.d/`.

#### SDK/CLI

- Are there new checks included in this PR? No

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

## Summary by CodeRabbit

* **New Features**
  * Jira finding delivery now returns detailed outcomes, including issue keys, IDs, URLs, and delivery status.
  * Added bulk issue-status checks covering open, completed, moved, missing, restricted, and unknown issues.
  * Added delivery-attempt search to help locate previously submitted findings.
  * Finding and delivery labels are now safely sanitized and length-limited.
  * Jira issue links are available directly from the integration.

* **Bug Fixes**
  * Improved handling and reporting of authentication, HTTP, and uncertain delivery failures.
  * Jira configuration and destination validation results are cached for more consistent behavior.

### CODE REVIEW

File: prowler/lib/outputs/jira/jira.py [hunk 1]
```diff
@@ -2195,6 +2248,84 @@ def _row(cells: list[dict]) -> dict:
...
+            JiraGetIssuesStatusResponseError,
+        ) as error:
+            raise error
+        except Exception as e:
+            error_message = f"Failed to get issues status from Jira: {e}"
+            logger.error(error_message)
+            raise JiraGetIssuesStatusError(
+                message=error_message, file=os.path.basename(__file__)
+            )
+
```
[the remark this answers is not in the document]
Comment (danibarranqueroo) [reply prowler/lib/outputs/jira/jira.py]: Fixed in `27bd08b072`: set-based dedup (order kept). The all-or-nothing behaviour is intentional and now documented in the docstring — callers treat keys missing from the result as deleted issues, so a partial result after a transient 429/502 would make live issues look gone and the API would create duplicates. Added `test_get_issues_status_is_all_or_nothing` for it.

File: tests/lib/outputs/jira/jira_test.py [hunk 2]
```diff
@@ -2003,9 +2004,271 @@ def test_send_finding_successful(
...
+        sizes = [
+            len(call.kwargs["json"]["issueIdsOrKeys"])
+            for call in mock_post.call_args_list
+        ]
+        assert sizes == [100, 100, 50]
+
+    def test_get_issues_status_without_keys(self):
+        """Test that no request is made when there is nothing to look up."""
+        assert self.jira_integration.get_issues_status([]) == {}
+        assert self.jira_integration.get_issues_status(["nope"]) == {}
```
[the remark this answers is not in the document]
Comment (danibarranqueroo) [reply tests/lib/outputs/jira/jira_test.py]: Fixed in `27bd08b072`: `requests.post` is patched and asserted not called.

**Comment by danibarranqueroo** [comment]
### Manual verification against a real Jira Cloud site (internal sandbox project)

Run from this branch with basic auth (`Jira(user_mail=…, api_token=…, domain=…)`). Site, project key and issue ids redacted.

1. `Jira.sanitize_labels(["prowler", "with space", "prowler", "", "ctrl\x00char", "x"*300])` → `['prowler', 'with_space', 'ctrlchar', 'xxx…']` (last one truncated to 255).
2. `jira.site_url` → `https://<site>.atlassian.net` (basic auth, derived from the domain).
3. `send_finding(..., issue_labels=[…, "prowler-finding-prowler-aws-s3_bucket_public_access-123456789012-us-east-1-my bucket/test"], finding_url=…, tenant_info=…)` → returned `{"key": "<PROJECT>-N", "id": "<id>", "url": "https://<site>.atlassian.net/browse/<PROJECT>-N"}`.
   The issue carries the labels `prowler`, `prowler-aws`, `prowler-high`, `prowler-s3_bucket_public_access`, `prowler-finding-…-my_bucket/test` (space sanitized to `_`), plus the **Finding URL** and **Tenant Info** rows in the description.
4. `get_issues_status(["<PROJECT>-N", "<PROJECT>-999999", "not a key"])` → `{"<PROJECT>-N": {"id": "<id>", "status": "To Do", "status_category": "new"}}` — non-existent and malformed keys are simply absent.
5. Transitioned the issue to Done in Jira → `get_issues_status(["<PROJECT>-N"])` → `{"<PROJECT>-N": {..., "status": "Done", "status_category": "done"}}`.
6. 250 keys (249 bogus + the real one) → 3 bulk-fetch requests, result contains only the real key.

Test issue closed afterwards.

File: prowler/lib/outputs/jira/jira.py [hunk 3]
```diff
@@ -2195,6 +2248,89 @@ def _row(cells: list[dict]) -> dict:
...
+            - JiraNoTokenError: Failed to get an access token
+            - JiraGetIssuesStatusResponseError: Jira returned a non-200 response
+            - JiraGetIssuesStatusError: Failed to get the issues status
+        """
+        valid_keys: list[str] = []
+        seen_keys: set[str] = set()
+        for key in issue_keys or []:
+            if key and self.ISSUE_KEY_REGEX.match(str(key)) and key not in seen_keys:
+                seen_keys.add(key)
+                valid_keys.append(key)
```
[the remark this answers is not in the document]
Comment (danibarranqueroo) [reply prowler/lib/outputs/jira/jira.py]: Fixed in `6c3c5f82bc`: `fullmatch` instead of `match`, with regression cases for `"SEC-1\n"` and keys with surrounding whitespace (no request is made for them).

File: prowler/lib/outputs/jira/jira.py [hunk 4]
```diff
@@ -2442,43 +3074,43 @@
...
+                    delivery_marker=delivery_attempt_marker,
+                    error_code="missing_credentials",
+                    error_message="Jira credentials are unavailable.",
                 )
 
-            available_issue_types = self.get_available_issue_types(project_key)
+            destination = (project_key, issue_type)
+            if destination not in self._validated_destinations:
+                projects = self.get_projects()
+                if project_key not in projects:
```
[the remark this answers is not in the document]
Comment (josema-xyz) [reply prowler/lib/outputs/jira/jira.py:3118]: Fixed. `get_projects()` now raises on a missing access token, so the membership check only receives a project mapping.

File: prowler/lib/outputs/jira/jira.py [hunk 5]
```diff
@@ -2442,43 +3074,43 @@
...
+                        error_message="The Jira project key is invalid.",
+                    )
 
-            if issue_type not in available_issue_types:
-                logger.error("The issue type is invalid")
-                raise JiraInvalidIssueTypeError(
-                    message="The issue type is invalid", file=os.path.basename(__file__)
-                )
+                available_issue_types = self.get_available_issue_types(project_key)
+                if issue_type not in available_issue_types:
```
[the remark this answers is not in the document]
Comment (josema-xyz) [reply prowler/lib/outputs/jira/jira.py:3128]: Fixed. `get_available_issue_types()` now raises on a missing access token, so the membership check only receives a list of issue types.

File: api/src/backend/tasks/tests/test_integrations.py [hunk 6]
```diff
@@ -20,6 +21,11 @@
     upload_security_hub_integration,
 )
 
+CONFIRMED_JIRA_CREATION = SimpleNamespace(
+    is_confirmed_success=True,
+    error_message=None,
+)
```
[the remark this answers is not in the document]
Comment (josema-xyz) [reply api/src/backend/tasks/tests/test_integrations.py]: Removed this API test change from this SDK-only PR. The API keeps `if result:` until #12541 adopts `JiraCreationResult` after `api/uv.lock` is updated.

File: prowler/lib/outputs/jira/jira.py [hunk 7]
```diff
@@ -432,6 +452,255 @@ def _sanitize_summary(summary: str) -> str:
...
+                    ),
+                    **response_details,
+                )
+            return JiraCreationResult(
+                outcome=JiraCreationOutcome.CONFIRMED_SUCCESS,
+                issue_key=issue_key,
+                issue_id=issue_id,
+                issue_url=issue_url,
+                **response_details,
+            )
```
[the remark this answers is not in the document]
Comment (josema-xyz) [reply prowler/lib/outputs/jira/jira.py:615]: Keeping this as designed. `confirmed_success` requires a usable issue key, immutable ID, and browse URL. A `201` without a derived URL remains `uncertain` and is reconciled through the delivery marker.

File: prowler/lib/outputs/jira/jira.py [hunk 8]
```diff
@@ -2195,12 +2472,363 @@ def _row(cells: list[dict]) -> dict:
...
+                            issue_id=issue_id,
+                            issue_key=issue_key,
+                            issue_url=issue_url,
+                        )
+                    )
+
+            next_page_token = response_json.get("nextPageToken")
+            if not next_page_token:
+                return result(JiraIssueSearchOutcome.SUCCESS, response=response)
+            payload["nextPageToken"] = next_page_token
```
[the remark this answers is not in the document]
Comment (josema-xyz) [reply prowler/lib/outputs/jira/jira.py:2847]: Fixed. The lookup tracks each `nextPageToken` and returns `unknown` with `pagination_stalled` when Jira repeats a token. It does not report an incomplete search as successful.

File: prowler/lib/outputs/jira/jira.py [hunk 9]
```diff
@@ -2442,43 +3074,43 @@ def send_finding(
...
+                available_issue_types = self.get_available_issue_types(project_key)
+                if issue_type not in available_issue_types:
+                    logger.error("The issue type is invalid")
+                    return JiraCreationResult(
+                        outcome=JiraCreationOutcome.CONFIRMED_REJECTION,
+                        delivery_marker=delivery_attempt_marker,
+                        error_code="invalid_issue_type",
+                        error_message="The Jira issue type is invalid.",
+                    )
+                self._validated_destinations.add(destination)
```
[the remark this answers is not in the document]
Comment (josema-xyz) [reply prowler/lib/outputs/jira/jira.py:3136]: The concrete failure paths are fixed. `get_projects()` and `get_available_issue_types()` now raise instead of returning exception objects, and `JiraGetAccessTokenError` maps to `retryable_failure`. `JiraGetCloudIDError` is not raised by `send_finding()`, so a blanket `except Exception` was not added because it would misclassify failures.

File: prowler/lib/outputs/jira/jira.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (josema-xyz) [reply prowler/lib/outputs/jira/jira.py]: This behavior changed after the review. `get_issues_status()` now preserves successful batches and returns `unknown` for references in a failed batch. Omitted issues are also `unknown` and are never treated as deleted.

**Comment by josema-xyz** [comment]
Correction to the earlier manual verification: it covers the previous key-based dictionary contract and should not be used as evidence for the current head.

The current implementation returns typed creation, marker-search, and issue-status results. Status lookups use immutable Jira issue IDs and report explicit `missing`, `forbidden`, and `unknown` outcomes. Missing or omitted issues are never treated as deleted.

The PR description and automated tests now reflect the current contract. Updated real-site verification will be posted after it is rerun against the final implementation.

File: prowler/lib/outputs/jira/jira.py [hunk 10]
```diff
@@ -2195,12 +2475,375 @@ def _row(cells: list[dict]) -> dict:
...
+                    error_code="invalid_reference",
+                    error_message="The stored Jira issue reference is invalid.",
+                )
+            else:
+                valid_references.append(reference)
+
+        if valid_references:
+            try:
+                access_token = self.get_access_token()
+            except (JiraRefreshTokenError, JiraRefreshTokenResponseError):
```
Comment (HugoPBrito) [inline prowler/lib/outputs/jira/jira.py]: I believe we should also catch `JiraGetAccessTokenError` here. `get_access_token()` can raise it, and `send_finding()` already treats it as an expected authentication failure.

Otherwise, this method raises instead of returning the documented `UNKNOWN` outcome. The same applies to `search_issues_by_delivery_attempt()` below. Can we handle it in both methods and add regression tests for each lookup path?

**Review by HugoPBrito — CHANGES_REQUESTED** [review]
Thanks for the thorough work here. I found one authentication edge case that can break the typed lookup contract, so I believe we should address it before merging.

File: prowler/lib/outputs/jira/jira.py (hunk 10 above)
Comment (josema-xyz) [reply prowler/lib/outputs/jira/jira.py]: Fixed. Both `get_issues_status()` and `search_issues_by_delivery_attempt()` now catch `JiraGetAccessTokenError` and return typed `unknown` results with `authentication_failed`. Regression tests cover both paths and verify that no Jira request is sent when token retrieval fails.

Comment (HugoPBrito) [reply prowler/lib/outputs/jira/jira.py]: Thanks!

**Review by HugoPBrito — APPROVED** [review]
